File: templates/Virtual-Assistant-Template/typescript/generator-botbuilder-assistant/generators/app/templates/customAssistant/src/dialogs/escalateDialog.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { BotTelemetryClient } from 'botbuilder';
import {
    ComponentDialog,
    DialogTurnResult,
    WaterfallDialog,
    WaterfallStepContext } from 'botbuilder-dialogs';
import { EscalateResponses } from '../responses/escalateResponses';
import { BotServices } from '../services/botServices';

export class EscalateDialog extends ComponentDialog {
    // Fields
    private readonly responder: EscalateResponses = new EscalateResponses();

    // Constructor
    constructor(botServices: BotServices, telemetryClient: BotTelemetryClient) {
        super(EscalateDialog.name);
        this.initialDialogId = EscalateDialog.name;
        const escalate: ((sc: WaterfallStepContext) => Promise<DialogTurnResult>)[] = [
            this.sendPhone.bind(this)
        ];
        this.addDialog(new WaterfallDialog(
            this.initialDialogId,
            escalate
        ));
    }

    private async sendPhone(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        await this.responder.replyWith(sc.context, EscalateResponses.responseIds.sendPhoneMessage);

        return sc.endDialog();
    }
}

File: samples/solutions/EnterpriseNotifications/VirtualAssistant/Dialogs/EscalateDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using VirtualAssistant.Responses.Escalate;
using VirtualAssistant.Services;

namespace VirtualAssistant.Dialogs
{
    public class EscalateDialog : ComponentDialog
    {
        private EscalateResponses _responder = new EscalateResponses();

        public EscalateDialog(BotServices botServices, IBotTelemetryClient telemetryClient)
            : base(nameof(EscalateDialog))
        {
            InitialDialogId = nameof(EscalateDialog);

            var escalate = new WaterfallStep[]
            {
                SendPhone,
            };

            AddDialog(new WaterfallDialog(InitialDialogId, escalate));
        }

        private async Task<DialogTurnResult> SendPhone(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            await _responder.ReplyWith(sc.Context, EscalateResponses.ResponseIds.SendPhoneMessage);
            return await sc.EndDialogAsync();
        }
    }
}
